python/src/scipp/plot/plot2d.py

```diff
@@ -0,0 +1,144 @@
+# @author Neil Vaytet
+
+# Scipp imports
+from .. import config
+from .controller2d import PlotController2d
+from .model2d import PlotModel2d
+from .profile import ProfileView
+from .sciplot import SciPlot
+from .view2d import PlotView2d
+
+
+def plot2d(scipp_obj_dict=None,
+           axes=None,
+           masks=None,
+           filename=None,
+           figsize=None,
+           ax=None,
+           cax=None,
+           aspect=None,
+           cmap=None,
+           log=False,
+           vmin=None,
+           vmax=None,
+           color=None,
+           logx=False,
+           logy=False,
+           logxy=False,
+           resolution=None):
+    """
+    Plot a 2D slice through a N dimensional dataset. For every dimension above
+    2, a slider is created to adjust the position of the slice in that
+    particular dimension.
+    """
+
+    sp = SciPlot2d(scipp_obj_dict=scipp_obj_dict,
+                   axes=axes,
+                   masks=masks,
+                   ax=ax,
+                   cax=cax,
+                   figsize=figsize,
+                   aspect=aspect,
+                   cmap=cmap,
+                   log=log,
+                   vmin=vmin,
+                   vmax=vmax,
+                   color=color,
+                   logx=logx or logxy,
+                   logy=logy or logxy,
+                   resolution=resolution)
+
+    if filename is not None:
+        sp.savefig(filename)
+
+    return sp
+
+
+class SciPlot2d(SciPlot):
+    def __init__(self,
+                 scipp_obj_dict=None,
+                 axes=None,
+                 masks=None,
+                 ax=None,
+                 cax=None,
+                 figsize=None,
+                 pax=None,
+                 aspect=None,
+                 cmap=None,
+                 log=None,
+                 vmin=None,
+                 vmax=None,
+                 color=None,
+                 logx=False,
+                 logy=False,
+                 resolution=None):
+
+        super().__init__()
+
+        # The main controller module which contains the slider widgets
+        self.controller = PlotController2d(scipp_obj_dict=scipp_obj_dict,
+                                           axes=axes,
+                                           masks=masks,
+                                           cmap=cmap,
+                                           log=log,
+                                           vmin=vmin,
+                                           vmax=vmax,
+                                           color=color,
+                                           logx=logx,
+                                           logy=logy,
+                                           button_options=['X', 'Y'])
+
+        # The model which takes care of all heavy calculations
+        self.model = PlotModel2d(controller=self.controller,
+                                 scipp_obj_dict=scipp_obj_dict,
+                                 resolution=resolution)
```

Comment:
So something feels backwards here: Passing the controller to the model (and the view) looks odd, in particular since the model implementation appears to be extracting data/information from the controller? Isn't that the coupling we would like to avoid with this design? I probably have an incomplete picture, but my interpretation of the MVC pattern would be that the controller should depend on the interfaces of the model and the view. Concrete models and views could then, e.g., be constructed independently and passed to the controller constructor. The controller would then take care of calling potentially required model/view methods for initializing them.

Reply:
I agree, this part is what still doesn't feel quite right, and also because of what you describe just below.
However, if i pass the `view` to the `controller`, how do i establish (i.e. what syntax would you suggest) the connection between the `view` and the `controller`.

It would be straightforward for the `controller` to send changes to the `view`, but how would the `view` send changes back to the controller, in the event of, for instance, zooming in on a 2d image where the `controller` has to initiate a resampling operation?

Reply:
One possibility for how the view communicates with the controller would be to implement some variation of the Observer/Subscriber design pattern. 

The full pattern would involve two interfaces where the controller implements an interface containing functions to react to changes in the view so in the example above might have a resample on zoom method. And the view has an interface to let the controller subscribe to changes.

In python I find that creating interfaces can be a bit clunky however and isn;t really well supported by the language and the fact that you can easilt pass around functions removes much of the need for it. So you can instead have the controller passing the view call_back functions which the view then calls when certain events happen. 

So as a concrete example you inject the view into the controller and then have the controller register the required callbacks with the view.

Reply:
As Matthew says, you can pass functions, or you can use a defined interface, without necessarily having it as a base class (since this is not required in Python). The controller coul just call

```python
view.set_controller(self)
```
Then the view can use
```
controller.set_zoom(params)
```
Might be that just passing the callbacks is cleaner though, i.e., what Matthew suggest.